stackrox/stackrox
# ROX-8789: Change operator index to file based catalog

## Description

Switching to file format for operator index. Following changes are made:
- bash utils functions are introduced. This can be used later also in Makefile where necessary.
- bash logic is moved from Makefile to custom bash script
- switched to file format for operator index

**Notes**

We are using `yaml` format for the index catalog. The main reason is that in-place manipulation of the YAML file is possible with `yq`. And in-place manipulation is problematic with `jq`. `yq` command is fetched similarly as `opm` -> download from GitHub.

## Checklist
- ~~[ ] Unit test and regression tests added~~ - there are no new tests here, since we are changing build scripts
- ~~[ ] Determined and documented upgrade steps~~ - no required update steps for clients
- ~~[ ] Documented user facing changes (create PR based on [stackrox/openshift-docs](https://github.com/stackrox/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~ - no user facing changes

## To clarify

- It's not clear if this change is relevant for CHANGELOG?

## Testing Performed

1. We need a local registry because `opm` requires it
```
docker run -d -p 5000:5000 --restart=always --name registry registry:2
```

2. Build a bundle and push it to the local registry
```
make bundle-build
docker tag stackrox/stackrox-operator-bundle:v3.69.0-396-g0a7127af42 localhost:5000/stackrox/stackrox-operator-bundle:v3.69.0-396-g0a7127af42
docker push localhost:5000/stackrox/stackrox-operator-bundle:v3.69.0-396-g0a7127af42
```

3. Build index. We can use the script directly. We need the `--use-http` flag because of `localhost:5000`. It should work without the `-use-http` flag on an existing build platform.
```
cd operator
./hack/make-index-build-step.sh --base-index-tag docker.io/stackrox/stackrox-operator-index:v3.68.0 --replaced-version v3.68.0 --bundle-tag localhost:5000/stackrox/stackrox-operator-bundle:v3.69.0-396-g0a7127af42 --index-tag localhost:5000/stackrox/stackrox-operator-index:v3.69.0-396-g0a7127af42 --clean-build --use-http
```

4. Compare old and new index

- Testing old index
```
docker run -d -p 50051:50051 docker.io/stackrox/stackrox-operator-index:v3.68.0
grpcurl -plaintext -d '{"channelName":"latest","csvName":"rhacs-operator.v3.67.0","pkgName":"rhacs-operator"}' localhost:50051 api.Registry/GetBundleThatReplaces
# There should be some result
grpcurl -plaintext -d '{"channelName":"latest","csvName":"rhacs-operator.v3.68.0","pkgName":"rhacs-operator"}' localhost:50051 api.Registry/GetBundleThatReplaces
# We should get an error: "no entry found for rhacs-operator latest"
```

- Testing new index (on another port - 50052)
```
docker run -d -p 50052:50051 localhost:5000/stackrox/stackrox-operator-index:v3.69.0-396-g0a7127af42
grpcurl -plaintext -d '{"channelName":"latest","csvName":"rhacs-operator.v3.67.0","pkgName":"rhacs-operator"}' localhost:50052 api.Registry/GetBundleThatReplaces
# There should be some result
grpcurl -plaintext -d '{"channelName":"latest","csvName":"rhacs-operator.v3.68.0","pkgName":"rhacs-operator"}' localhost:50052 api.Registry/GetBundleThatReplaces
# We should also get result, because we have replacement for v3.68.0 in new catalog
```
